### PR TITLE
Fix valiate_index and validate_dataframe for indices that don't start with `0`

### PR DESCRIPTION
**Pull request by Kully**
For the development of this FF: https://github.com/plotly/plotly.py/issues/897
ready for a review
(changelog updated)
cc @cldougl @chriddyp

### CODE REVIEW

**Comment by chriddyp** [comment]
@Kully  Could you please add a suite of tests for this behaviour? Since this PR includes a "fix" there should be tests that display how before this functionality was added the code was broken and after this functionality was added the code was fixed. You can even rebase your PR to add the tests at the beginning of your commits to show how the tests fail and then how your code fixes them.

**Comment by Kully** [comment]
> Since this PR includes a "fix"

I think I'm using the term "fix" incorrectly then. I actually don't think there were any tests for this particular function, but it was used implicitly through other functions and was luckily avoiding errors

**Comment by chriddyp** [comment]
> and was luckily avoiding errors

Then we need to illustrate a case where this code _would_ cause an error and the show how this code fixed that error.

All of our code and all of our code's pathways need to be tested, without exception.

**Comment by chriddyp** [comment]
> and was luckily avoiding errors

Then we need to illustrate a case where this code _would_ cause an error and the show how this code fixed that error.

All of our code and all of our code's pathways need to be tested, without exception. [Untested code is broken code](https://www.slideshare.net/wooda/philipp-von-weitershausen-untested-code-is-broken-code)

**Comment by Kully** [comment]
I want:
- `plotly.figure_factory.utils.validate_index` to not fail when it accepts a pandas Series with index not starting at `0`

ex: this fails
```
import plotly
import plotly.figure_factory.utils as utils
import pandas as pd

s = pd.Series([1, 2, 3], index=[7,7,7])

utils.validate_index(s)
```
with the error
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-10-b78ed750e2cf> in <module>()
      7 df = pd.DataFrame([[1, 2], [3, 4]], index=[1, 2])
      8 
----> 9 utils.validate_index(s)

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/plotly/figure_factory/utils.pyc in validate_index(index_vals)
     42     """
     43     from numbers import Number
---> 44     if isinstance(index_vals[0], Number):
     45         if not all(isinstance(item, Number) for item in index_vals):
     46             raise exceptions.PlotlyError("Error in indexing column. "

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/core/series.pyc in __getitem__(self, key)
    621         key = com._apply_if_callable(key, self)
    622         try:
--> 623             result = self.index.get_value(self, key)
    624 
    625             if not is_scalar(result):

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/core/indexes/base.pyc in get_value(self, series, key)
   2555         try:
   2556             return self._engine.get_value(s, k,
-> 2557                                           tz=getattr(series.dtype, 'tz', None))
   2558         except KeyError as e1:
   2559             if len(self) > 0 and self.inferred_type in ['integer', 'boolean']:

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_value()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_value()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine._get_loc_duplicates()

KeyError: 0
```

`plotly.figure_factory.utils.validate_dataframe`  to not fail when it accepts a pandas.Dataframe (current only works for lists of lists-type objects)

ex: this fails
```
import plotly
import plotly.figure_factory.utils as utils
import pandas as pd

s = pd.Series([1, 2, 3], index=[7,7,7])

df = pd.DataFrame([[1, 2], [3, 4]], index=[1, 2])

utils.validate_dataframe(df)
```
with error
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-8acd82df1c54> in <module>()
      7 df = pd.DataFrame([[1, 2], [3, 4]], index=[1, 2])
      8 
----> 9 utils.validate_dataframe(df)

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/plotly/figure_factory/utils.pyc in validate_dataframe(array)
     66     from numbers import Number
     67     for vector in array:
---> 68         if isinstance(vector[0], Number):
     69             if not all(isinstance(item, Number) for item in vector):
     70                 raise exceptions.PlotlyError("Error in dataframe. "

TypeError: 'int' object has no attribute '__getitem__'
```

**Comment by jonmmease** [comment]
Hi @Kully , is there anything in this PR that we should pick back up?

**Comment by Kully** [comment]
> , is there anything in this PR that we should pick back up?

Yeah  it's still an unresolved problem. I haven't worked on this PR in quite a while though. I'd be glad to merge in master and fix the conflicting files to get ready to merge

**Comment by Kully** [comment]
@jonmmease This PR is ready for a review now.

There are more improvements we can make to this function (i.e. name change to be more clear) but we should save that for another PR. (I'll open a new issue for that)

Let me know what you think of the if..else logic - it is not 100% exhaustive because these utils functions are designed to be used by figure factories which already expect list/tuple inputs. And with the validations on plotly objects now I think this should be fine.

File: plotly/figure_factory/utils.py [hunk 1]
```diff
@@ -47,20 +49,26 @@ def validate_index(index_vals):
...
+    if isinstance(first_item, Number):
         if not all(isinstance(item, Number) for item in index_vals):
-            raise exceptions.PlotlyError("Error in indexing column. "
-                                         "Make sure all entries of each "
-                                         "column are all numbers or "
-                                         "all strings.")
+            error = True
 
-    elif isinstance(index_vals[0], str):
+    elif isinstance(first_item, str):
```
[the remark this answers is not in the document]
Comment (Kully) [reply plotly/figure_factory/utils.py]: agreed

**Comment by Kully** [comment]
> Do you have any examples of figure factory calls that resulted in incorrect errors before and that plot properly with these updates?

You know, I thought there was an example I ran a while ago that resulted in a keyerror because of a bad index, but I am unable to reproduce.

I looked [https://github.com/plotly/plotly.py/blob/master/plotly/figure_factory/_scatterplot.py#L1103-L1112](here) at the legacy scatterplot module and see that each column is converted to a list before entering `validate_index`

**Comment by jonmmease** [comment]
Ok, I'm inclined to hold off until we run into an issue at the figure factory level. I haven't really dug into most of the figure factories enough yet to be confident I understand the downstream consequences of these changes.  What do you think?

**Comment by Kully** [comment]
> I'm inclined to hold off until we run into an issue at the figure factory level.

I'm good with that. Let's close this one for the time being.

**Pull request closed without merging**